### PR TITLE
chore: update package information report

### PR DESCRIPTION
**Pull request by JoeWang1127**
In this PR:
- Generate package information report for package that has no non-compliant licenses and security vulnerabilities.
- Encode version using `URLEncoder`. The version may contain special characters that needs to be encoded before sending to deps.dev, e.g., [9+181-r4173-1](https://deps.dev/maven/com.google.errorprone%3Ajavac-shaded/9%2B181-r4173-1).

An example of the report:
```
## Package information of io.opentelemetry:opentelemetry-api:1.37.0
Licenses: [Apache-2.0]
Vulnerabilities: None.
Checked in [deps.dev query](https://api.deps.dev/v3/query?versionKey.system=MAVEN&versionKey.name=io.opentelemetry:opentelemetry-api&versionKey.version=1.37.0)

## Dependencies:
## Package information of io.opentelemetry:opentelemetry-context:1.37.0
Licenses: [Apache-2.0]
Vulnerabilities: None.
Checked in [deps.dev query](https://api.deps.dev/v3/query?versionKey.system=MAVEN&versionKey.name=io.opentelemetry:opentelemetry-context&versionKey.version=1.37.0)
```

### CODE REVIEW

**Comment by JoeWang1127** [comment]
"Downstream Unmanaged Dependency Check / validate (java-bigtable)" is irrelevant to this PR.

File: java-shared-dependencies/dependency-analyzer/src/main/java/com/google/cloud/model/AnalysisResult.java [hunk 1]
```diff
@@ -92,11 +98,49 @@ private <T> void formatLog(VersionKey root, Map<VersionKey, List<T>> map, String
...
+      for (int i = 1; i < packageInfos.size(); i++) {
+        appendToReport(builder, packageInfos.get(i));
+      }
+    }
+    builder.append("\n\n");
+
+    return builder.toString();
+  }
+
+  private void appendToReport(StringBuilder builder, PackageInfo packageInfo) {
```
Comment (suztomo) [inline java-shared-dependencies/dependency-analyzer/src/main/java/com/google/cloud/model/AnalysisResult.java:129]: static?

File: java-shared-dependencies/dependency-analyzer/src/main/java/com/google/cloud/model/AnalysisResult.java [hunk 2]
```diff
@@ -92,11 +98,49 @@ private <T> void formatLog(VersionKey root, Map<VersionKey, List<T>> map, String
...
+    String relation = versionKey.equals(root) ? "self" : "dependency";
+    return String.format("====================== %s, %s of %s ======================",
+        versionKey, relation, root);
   }
 
   private String endSeparator() {
     return "===========================================================";
   }
+
+  private String packageInfoReport() {
```
Comment (suztomo) [inline java-shared-dependencies/dependency-analyzer/src/main/java/com/google/cloud/model/AnalysisResult.java:110]: Would you add a test case of the pull request description?


Java 17's test block should make it easy. https://blogs.oracle.com/javamagazine/post/text-blocks-come-to-java

File: java-shared-dependencies/dependency-analyzer/src/main/java/com/google/cloud/model/AnalysisResult.java [hunk 3]
```diff
@@ -92,11 +98,49 @@ private <T> void formatLog(VersionKey root, Map<VersionKey, List<T>> map, String
...
+    return builder.toString();
+  }
+
+  private void appendToReport(StringBuilder builder, PackageInfo packageInfo) {
+    VersionKey versionKey = packageInfo.versionKey();
+    // generate the report using Markdown format.
+    builder.append(String.format("## Package information of %s\n", versionKey));
+    builder.append(String.format("Licenses: %s\n", packageInfo.licenses()));
+    builder.append(
+        String.format("Vulnerabilities: None.\nChecked in [deps.dev query](%s)\n",
```
Comment (suztomo) [inline java-shared-dependencies/dependency-analyzer/src/main/java/com/google/cloud/model/AnalysisResult.java:135]: The visible part of the link should have the artifact coordinates.

File: java-shared-dependencies/dependency-analyzer/src/main/java/com/google/cloud/model/AnalysisResult.java [hunk 4]
```diff
@@ -92,11 +98,49 @@ private <T> void formatLog(VersionKey root, Map<VersionKey, List<T>> map, String
...
     return "===========================================================";
   }
+
+  private String packageInfoReport() {
+    StringBuilder builder = new StringBuilder();
+    builder.append("Please copy and paste the package information below to your ticket.\n");
+    builder.append("\n\n");
+    appendToReport(builder, packageInfos.get(0));
+
+    builder.append("## Dependencies:\n");
```
Comment (suztomo) [inline java-shared-dependencies/dependency-analyzer/src/main/java/com/google/cloud/model/AnalysisResult.java:116]: Having the headings of the same level line by line is strange. This means the 1st section is empty. Would you fix this somehow?

image

File: java-shared-dependencies/dependency-analyzer/src/main/java/com/google/cloud/model/AnalysisResult.java (hunk 3 above)
Comment (JoeWang1127) [reply java-shared-dependencies/dependency-analyzer/src/main/java/com/google/cloud/model/AnalysisResult.java]: Done.

File: java-shared-dependencies/dependency-analyzer/src/main/java/com/google/cloud/model/AnalysisResult.java (hunk 4 above)
Comment (JoeWang1127) [reply java-shared-dependencies/dependency-analyzer/src/main/java/com/google/cloud/model/AnalysisResult.java:116]: Done.

File: java-shared-dependencies/dependency-analyzer/src/main/java/com/google/cloud/model/AnalysisResult.java (hunk 2 above)
Comment (JoeWang1127) [reply java-shared-dependencies/dependency-analyzer/src/main/java/com/google/cloud/model/AnalysisResult.java:110]: > Would you add a test case of the pull request description?

It's not easily test using unit test as the encoded string needs to be sent to deps.dev to verify the correctness.

I'll write an integration test for it.

Comment (suztomo) [reply java-shared-dependencies/dependency-analyzer/src/main/java/com/google/cloud/model/AnalysisResult.java:110]: Would you write this in a way that the formatter function should operate on a Record class to generate the formatted string.

In this design, the unit test for the formatter does not need to access deps.dev.

**Review by suztomo — APPROVED** [review]
Approving this, as it's good enhancement already. We may add more pull requests later to follow up.

File: java-shared-dependencies/dependency-analyzer/src/main/java/com/google/cloud/model/AnalysisResult.java (hunk 2 above)
Comment (JoeWang1127) [reply java-shared-dependencies/dependency-analyzer/src/main/java/com/google/cloud/model/AnalysisResult.java:110]: Wrote unit tests to verify the encoded string.

**Pull request merged**